smaug/python/graph.py

```diff
@@ -75,7 +78,7 @@ def add_node(
 
     # If we are currently within a control flow context, add the new node to
     # the context.
-    if self._control_flow_context is not None:
+    if self._control_flow_context is not None and needs_control_flow_processing:
```

Comment:
This cyclic dependency is still here?

Reply:
For a switch node added by the control flow context, we set `needs_control_flow_processing` to False, so it doesn't call control_flow_context.add_node().

If you are talking about the nodes added by executing the two callables:
```python
with true_context:
  true_fn()

with false_context:
  false_fn()
```
Because we can't get the nodes in the callable until we execute it, I can't think of a way that can avoid calling context.add_node() while adding a node to the graph.

Reply:
Alright, I really think that a subgraph is the right way to do this. The problem is what you described - you need to intercept the creation of all the operators and usage of all the tensors in order to insert the switch node at the right place. What you're doing with the `ControlFlowContext` is creating a mechanism outside the graph, but that still requires the hooks you added. Using a subgraph would let you capture everything you want without any additional context managers. It would also work recursively - you could put subgraphs inside subgraphs, etc, which required additional tracking about that implicit hierarchy in `ControlFlowContext`.

Thinking through how to implement this with subgraphs:

```python
def cond(predicate, true_fn, false_fn):
   with Graph() as subgraph:
      true_ops = true_fn()
      false_ops = false_fn()
      insert_switch_nodes_for_all_tensors(predicate, subgraph.tensors)
      add_merge_node(true_ops, false_ops)
      subgraph.parent_graph.merge(subgraph)
```
Potential issues:
1. Would the subgraph be able to see the tensors being used in `true_fn`? It'll be captured by the callable so the variable exists, but I wonder if creating the operators will cause a lookup in the subgraph (which will fail). If this is the case, you might be tempted to add all the tensors of parent graph to subgraph when creating subgraph, but then that'll result in switch nodes being added for all tensors of the parent graph too. We may need a recursive lookup through the graph tree if so.
2. Merging graphs - will need to make sure you add any new tensors created as well as update properties of existing tensors (if applicable).

Reply:
Regarding the potential issues:

1. One thing I want to clarify is we only need to insert switch nodes for the input tensors of the subgraph, rather than all the tensors of it. And in order to know whether a tensor is an input, we'll need to intercept graph.add_node() - when a new node is added, we check whether its inputs are from nodes already in the graph or from outside (the parent graph). This is how `ControlFlowContext` keeps track of the input tensors. I feel like we are moving the functionalities of `ControlFlowContext` to Graph.

2. Merging graphs basically sets tensors between the output nodes of the parent graph to the input nodes of the subgraph. So again we'll need to know the inputs of the graph.

Reply:
1. This sounds like a significant amount of additional complexity inside add_node, no? You still need hooks inside graph.add_node to do more work. I would prefer to contain as much of the complexity of handling correct switch node placement inside the `cond()` operator implementation itself, not somewhere else. We should avoid letting operator-specific handling infect common structures like `Graph` as much as possible.

2. Yes, we'll need to know the inputs of the graph, and that's possible to do. We just need to take stock of all the tensors used by any node in the subgraph. If it doesn't appear in the local subgraph workspace, then it's an input tensor.

There's one other implementation option I thought of last night. We can add an additional predication input to *all* operator types, which will cause the node to be marked as dead if the predicate is false. This may involve some more changes to the C++ side, but it wouldn't require any additional context managers, tracking of external tensors, etc.

Reply:
I see. So sounds like the solution for determining input tensors of the subgraph is do a graph-level pass of it and keep the tracking info maintained within `cond()`, right?

Adding a new input to every operator sounds like a pretty large change for solving a small issue. I prefer to contain the issue within the Python side as much as possible.

Reply:
Not necessarily a graph-traversal pass. Just scan over all the tensors used by all the operators. And yes, keep tracking info contained within cond().

Reply:
Sounds good. I'll do this then.